### PR TITLE
RUN 3772 queue up console.* messages from the app and send them to the RVM

### PR DESCRIPTION
**Pull request by patpao**
[JIRA](https://appoji.jira.com/browse/RUN-3772)

As part of log management, we want to send console.log, console.info, console.error, etc. messages from the core to the RVM, and the RVM will write them to separate files (to avoid writing logs from every application to the same debug.log). Sending a message from the core to the RVM on every console log is too hard on the browser process (and sends a lot of messages to the RVM), so instead we store all the messages in a queue. Every time a message is pushed to an empty queue, we set a timer to flush it in 10s.

Tests:
[Win7](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/5a7487ad3f32861e82b77d08)
[Win10](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/5a74881c3f32861e82b77d09)

### CODE REVIEW

File: src/browser/api/window.js [hunk 1]
```diff
@@ -837,6 +838,40 @@ Window.create = function(id, opts) {
...
+        const date = new Date();
+        const month = checkPrependLeadingZero(date.getMonth() + 1);
+        const day = checkPrependLeadingZero(date.getDate());
+        const year = String(date.getFullYear()).slice(2);
+        const hour = checkPrependLeadingZero(date.getHours());
+        const minute = checkPrependLeadingZero(date.getMinutes());
+        const second = checkPrependLeadingZero(date.getSeconds());
+
+        // Format timestamp to match debug.log
+        const timeStamp = '' + month + '/' + day + '/' + year + ' ' + hour + ':' + minute + ':' + second;
```
Comment (patpao) [inline src/browser/api/window.js]: Open to suggestions on this - couldn't find an easy way to format the time

Comment (whyn07m3) [reply src/browser/api/window.js]: I'd suggest ```const timeStamp = `${month}/${day}/${year} ${hour}:${minute}:${second}`;```

File: src/browser/api/window.js [hunk 2]
```diff
@@ -837,6 +838,40 @@ Window.create = function(id, opts) {
...
+
+        addConsoleMessageToRVMMessageQueue({
```
Comment (whyn07m3) [inline src/browser/api/window.js]: Can be simplified to:
```addConsoleMessageToRVMMessageQueue({  level, message, timestamp, appConfigUrl });```
For this change a constant on top needs to change its name

File: src/browser/rvm/rvm_message_bus.ts [hunk 3]
```diff
@@ -26,12 +26,30 @@ export interface RvmMsgBase {
...
 // topic: application -----
 type applicationTopic = 'application';
+type applicationLogAction = 'application-log';
 type registerUserAction = 'register-user';
 type hideSplashscreenAction = 'hide-splashscreen';
 type relaunchOnCloseAction = 'relaunch-on-close';
 type getDesktopOwnerSettingsAction = 'get-desktop-owner-settings';
 type downloadRuntimeAction = 'runtime-download';
 
+export interface ConsoleMessage {
```
Comment (whyn07m3) [inline src/browser/rvm/rvm_message_bus.ts]: Alphabetical order

File: src/browser/rvm/rvm_message_bus.ts [hunk 4]
```diff
@@ -26,12 +26,30 @@ export interface RvmMsgBase {
...
+    level: number;
+    message: string;
+    appConfigUrl: string;
+    timeStamp: string;
+}
+
+export interface ApplicationLog extends RvmMsgBase {
```
Comment (whyn07m3) [inline src/browser/rvm/rvm_message_bus.ts]: Alphabetical order

File: src/browser/rvm/utils.ts [hunk 5]
```diff
@@ -14,6 +15,42 @@ interface SendToRVMOpts {
...
     sourceUrl?: string;
     data?: any;
+    runtimeVersion?: string;
+    payload?: any;
+}
+
+let consoleMessageQueue: ConsoleMessage[] = [];
+let isFlushScheduled: boolean = false;
+
+function flushConsoleMessageQueue() {
```
Comment (whyn07m3) [inline src/browser/rvm/utils.ts]: `function flushConsoleMessageQueue(): void {`

**Comment by patpao** [comment]
Thanks for reviewing @whyn07m3 , addressed your comments

File: src/browser/rvm/rvm_message_bus.ts [hunk 6]
```diff
@@ -50,6 +43,13 @@ export interface ApplicationLog extends RvmMsgBase {
     };
 }
 
+export interface ConsoleMessage {
```
Comment (whyn07m3) [inline src/browser/rvm/rvm_message_bus.ts]: Alphabetical order please

Comment (patpao) [reply src/browser/rvm/rvm_message_bus.ts]: Fixed

**Comment by StevenEBarbaro** [comment]
🌳 Changes pending for additionally flushing the queue if a message size is reached

File: src/browser/api/window.js [hunk 7]
```diff
@@ -837,6 +838,48 @@ Window.create = function(id, opts) {
...
+        // if you just called "addConsoleMessageToRVMMessageQueue" directly from here. So to get around that, we
+        // wrap this entire function in a "setTimeout" to put it in a different context. Eventually we should figure
+        // out if there is a way around this by using event.preventDefault or something similar
+        setTimeout(() => {
+            electronApp.vlog(1, `CONSOLE-MESSAGE`);
+
+            const appConfigUrl = coreState.getConfigUrlByUuid(identity.uuid);
+            if (!appConfigUrl) {
+                electronApp.vlog(2, `Error: could not get manifest url for app with uuid: ${identity.uuid}`);
+            }
```
Comment (wenjunche) [inline src/browser/api/window.js]: if appConfigUrl is required, should return here ?

File: src/browser/rvm/utils.ts [hunk 8]
```diff
@@ -14,6 +16,41 @@ interface SendToRVMOpts {
...
+
+export function addConsoleMessageToRVMMessageQueue(message: ConsoleMessage): void {
+    consoleMessageQueue.push(message);
+
+    // If no timer already set, set one to flush the queue in 10s
+    if (!isFlushScheduled) {
+        isFlushScheduled = true;
+        setTimeout(flushConsoleMessageQueue, 10000);
+    }
 }
```
Comment (wenjunche) [inline src/browser/rvm/utils.ts]: force Flush if consoleMessageQueue.length > someThreshold ?

**Pull request merged**